File: cpp/samples/core/universalid.cpp
/**
 * @file universalid.cpp
 * @brief Demonstrates usage of vista-sdk UniversalId and UniversalIdBuilder APIs
 * @details This sample shows how to build UniversalIds combining IMO numbers with LocalIds,
 *          parse UniversalId strings, and work with globally unique sensor identifiers
 */

#include <dnv/VistaSDK.h>

#include <iomanip>
#include <iostream>
#include <string>
#include <vector>

int main()
{
    using namespace dnv::vista::sdk;

    std::cout << "=== vista-sdk UniversalId Sample ===\n\n";

    {
        std::cout << "1. UniversalIdBuilder: Building a simple UniversalId\n";
        std::cout << "------------------------------------------------------\n";

        const auto& vis = VIS::instance();
        const auto& gmod = vis.gmod(vis.latest());
        const auto& locations = vis.locations(vis.latest());
        const auto& codebooks = vis.codebooks(vis.latest());

        // Create IMO number
        auto imoNumber = ImoNumber{ 1234567 };

        // Create LocalIdBuilder
        auto primaryItem = GmodPath::fromShortPath("411.1/C101.31-2", gmod, locations);
        auto qtyTag = codebooks[CodebookName::Quantity].createTag("temperature");

        if (primaryItem.has_value() && qtyTag.has_value())
        {
            auto localIdBuilder =
                LocalIdBuilder::create(vis.latest()).withPrimaryItem(*primaryItem).withMetadataTag(*qtyTag);

            // Build UniversalId
            auto universalId =
                UniversalIdBuilder::create(vis.latest()).withImoNumber(imoNumber).withLocalId(localIdBuilder).build();

            std::cout << "Built UniversalId: " << universalId.toString() << "\n";
            std::cout << "  IMO Number: " << universalId.imoNumber().toString() << "\n";
            std::cout << "  LocalId   : " << universalId.localId().toString() << "\n";
        }

        std::cout << "\n";
    }

    {
        std::cout << "2. UniversalIdBuilder: Building with full LocalId\n";
        std::cout << "---------------------------------------------------\n";

        const auto& vis = VIS::instance();
        const auto& gmod = vis.gmod(vis.latest());
        const auto& locations = vis.locations(vis.latest());
        const auto& codebooks = vis.codebooks(vis.latest());

        // Build a complete LocalId with secondary item and multiple metadata tags
        auto primaryItem = GmodPath::fromShortPath("621.21/S90", gmod, locations);
        auto secondaryItem = GmodPath::fromShortPath("411.1/C101", gmod, locations);
        auto qtyTag = codebooks[CodebookName::Quantity].createTag("mass");
        auto cntTag = codebooks[CodebookName::Content].createTag("fuel.oil");
        auto posTag = codebooks[CodebookName::Position].createTag("inlet");

        if (primaryItem.has_value() && secondaryItem.has_value() && qtyTag.has_value() && cntTag.has_value() &&
            posTag.has_value())
        {
            auto localIdBuilder = LocalIdBuilder::create(vis.latest())
                                      .withPrimaryItem(*primaryItem)
                                      .withSecondaryItem(*secondaryItem)
                                      .withMetadataTag(*qtyTag)
                                      .withMetadataTag(*cntTag)
                                      .withMetadataTag(*posTag);

            // Create UniversalId
            auto imoNumber = ImoNumber{ "IMO9074729" };
            auto universalId =
                UniversalIdBuilder::create(vis.latest()).withImoNumber(imoNumber).withLocalId(localIdBuilder).build();

            std::cout << "Built UniversalId: " << universalId.toString() << "\n";
            std::cout << "  Format       : {naming-entity}/{imo-number}{local-id}\n";
            std::cout << "  Naming entity: " << UniversalId::namingEntity() << "\n";
            std::cout << "  IMO Number   : " << universalId.imoNumber().toString() << "\n";
        }

        std::cout << "\n";
    }

    {
        std::cout << "3. UniversalId: Parsing from string\n";
        std::cout << "-------------------------------------\n";

        std::string universalIdStr = "data.dnv.com/IMO1234567/dnv-v2/vis-3-4a/411.1/C101.31-2/meta/qty-temperature";

        // Parse without error details
        auto universalId = UniversalId::fromString(universalIdStr);

        if (universalId.has_value())
        {
            std::cout << "Parsed successfully: " << universalId->toString() << "\n";
            std::cout << "  IMO Number : " << universalId->imoNumber().toString() << "\n";
            std::cout << "  LocalId    : " << universalId->localId().toString() << "\n";
            std::cout << "  VIS Version: " << VisVersions::toString(universalId->localId().version()) << "\n";
        }
        else
        {
            std::cout << "Parse failed\n";
        }

        std::cout << "\n";
    }

    {
        std::cout << "4. UniversalId: Parsing with error handling\n";
        std::cout << "----------------------------------------------\n";

        std::vector<std::string> testStrings = {
            "data.dnv.com/IMO1234567/dnv-v2/vis-3-4a/411.1/C101.31-2/meta/qty-temperature",
            "data.dnv.com/IMO9074729/dnv-v2/vis-3-4a/621.21/S90/sec/411.1/C101/meta/qty-mass/cnt-fuel.oil/pos-inlet",
            "data.dnv.com/IMO9785811/dnv-v2/vis-3-7a/612.21/C701.23/C633/meta/calc~accumulate",
            "",
            "data.dnv.com/INVALID/dnv-v2/vis-3-4a/411.1/meta/qty-temperature",
            "wrong.entity/IMO1234567/dnv-v2/vis-3-4a/411.1/meta/qty-temperature",
            "data.dnv.com/IMO1234567/dnv-v2/vis-3-4a/INVALID/meta/qty-temperature"
        };

        for (const auto& str : testStrings)
        {
            ParsingErrors errors;
            auto universalId = UniversalId::fromString(str, errors);

            std::cout << "Parsing: \"" << (str.empty() ? "(empty)" : str.substr(0, 60))
                      << (str.length() > 60 ? "..." : "") << "\"\n";

            if (universalId.has_value())
            {
                std::cout << "  Success\n";
                std::cout << "    IMO: " << universalId->imoNumber().toString() << "\n";
            }
            else
            {
                std::cout << "  Failed:\n";
                if (errors.hasErrors())
                {
                    for (const auto& [type, message] : errors)
                    {
                        std::cout << "    - " << message << "\n";
                    }
                }
            }
            std::cout << "\n";
        }
    }

    {
        std::cout << "5. UniversalIdBuilder: Modifying existing UniversalIds\n";
        std::cout << "--------------------------------------------------------\n";

        auto original =
            UniversalId::fromString("data.dnv.com/IMO1234567/dnv-v2/vis-3-4a/411.1/C101.31-2/meta/qty-temperature");

        if (original.has_value())
        {
            std::cout << "Original                    : " << original->toString() << "\n";

            const auto& vis = VIS::instance();
            const auto& codebooks = vis.codebooks(vis.latest());

            // Change IMO number
            auto newImo = ImoNumber{ 9074729 };
            auto modified = original->builder().withImoNumber(newImo).build();

            std::cout << "Modified (new IMO)          : " << modified.toString() << "\n";

            // Modify the LocalId part
            auto cntTag = codebooks[CodebookName::Content].createTag("water");
            if (cntTag.has_value())
            {
                auto modified2 =
                    original->builder().withLocalId(original->localId().builder().withMetadataTag(*cntTag)).build();

                std::cout << "Modified (added content tag): " << modified2.toString() << "\n";
            }
        }

        std::cout << "\n";
    }

    {
        std::cout << "6. UniversalId: Accessing components\n";
        std::cout << "--------------------------------------\n";

        auto universalId = UniversalId::fromString(
            "data.dnv.com/IMO9074729/dnv-v2/vis-3-4a/621.21/S90/sec/411.1/C101/meta/qty-mass/cnt-fuel.oil/pos-inlet");

        if (universalId.has_value())
        {
            std::cout << "UniversalId: " << universalId->toString() << "\n\n";

            std::cout << "IMO Number:\n";
            std::cout << "  Value: " << universalId->imoNumber().toString() << "\n";

            std::cout << "\nLocalId components:\n";
            const auto& localId = universalId->localId();
            std::cout << "  VIS Version   : " << VisVersions::toString(localId.version()) << "\n";
            std::cout << "  Primary item  : " << localId.primaryItem().toString() << "\n";

            if (localId.secondaryItem().has_value())
            {
                std::cout << "  Secondary item: " << localId.secondaryItem()->toString() << "\n";
            }

            std::cout << "\nMetadata tags:\n";
            if (localId.quantity().has_value())
                std::cout << "  Quantity: " << localId.quantity()->value() << "\n";
            if (localId.content().has_value())
                std::cout << "  Content : " << localId.content()->value() << "\n";
            if (localId.position().has_value())
                std::cout << "  Position: " << localId.position()->value() << "\n";
        }

        std::cout << "\n";
    }

    {
        std::cout << "7. UniversalId: Equality and comparison\n";
        std::cout << "-----------------------------------------\n";

        auto uid1 =
            UniversalId::fromString("data.dnv.com/IMO1234567/dnv-v2/vis-3-4a/411.1/C101.31-2/meta/qty-temperature");
        auto uid2 =
            UniversalId::fromString("data.dnv.com/IMO1234567/dnv-v2/vis-3-4a/411.1/C101.31-2/meta/qty-temperature");
        auto uid3 =
            UniversalId::fromString("data.dnv.com/IMO9074729/dnv-v2/vis-3-4a/411.1/C101.31-2/meta/qty-temperature");
        auto uid4 =
            UniversalId::fromString("data.dnv.com/IMO1234567/dnv-v2/vis-3-4a/411.1/C101.31-2/meta/qty-pressure");

        std::cout << std::boolalpha;
        if (uid1.has_value() && uid2.has_value() && uid3.has_value() && uid4.has_value())
        {
            std::cout << "UniversalId1: Same IMO, same LocalId\n";
            std::cout << "UniversalId2: Same IMO, same LocalId\n";
            std::cout << "UniversalId3: Different IMO, same LocalId\n";
            std::cout << "UniversalId4: Same IMO, different LocalId\n\n";

            std::cout << "uid1 == uid2? " << (*uid1 == *uid2) << " (same IMO & LocalId)\n";
            std::cout << "uid1 == uid3? " << (*uid1 == *uid3) << " (different IMO)\n";
            std::cout << "uid1 == uid4? " << (*uid1 == *uid4) << " (different LocalId)\n";
            std::cout << "uid1 != uid3? " << (*uid1 != *uid3) << "\n";
        }

        std::cout << "\n";
    }

    {
        std::cout << "8. UniversalIdBuilder: Validation\n";
        std::cout << "-----------------------------------\n";

        const auto& vis = VIS::instance();
        const auto& gmod = vis.gmod(vis.latest());
        const auto& locations = vis.locations(vis.latest());
        const auto& codebooks = vis.codebooks(vis.latest());

        auto imoNumber = ImoNumber{ 1234567 };
        auto primaryItem = GmodPath::fromShortPath("411.1/C101.31-2", gmod, locations);
        auto qtyTag = codebooks[CodebookName::Quantity].createTag("temperature");

        if (primaryItem.has_value() && qtyTag.has_value())
        {
            auto localIdBuilder =
                LocalIdBuilder::create(vis.latest()).withPrimaryItem(*primaryItem).withMetadataTag(*qtyTag);

            // Valid builder
            auto validBuilder =
                UniversalIdBuilder::create(vis.latest()).withImoNumber(imoNumber).withLocalId(localIdBuilder);

            std::cout << std::boolalpha;
            std::cout << "Valid builder:\n";
            std::cout << "  Is valid: " << validBuilder.isValid() << "\n";

            // Missing IMO number
            auto missingImo = UniversalIdBuilder::create(vis.latest()).withLocalId(localIdBuilder);
            std::cout << "\nBuilder missing IMO number:\n";
            std::cout << "  Is valid: " << missingImo.isValid() << "\n";

            // Missing LocalId
            auto missingLocalId = UniversalIdBuilder::create(vis.latest()).withImoNumber(imoNumber);
            std::cout << "\nBuilder missing LocalId:\n";
            std::cout << "  Is valid: " << missingLocalId.isValid() << "\n";

            // Try to build invalid
            std::cout << "\nAttempting to build without IMO number:\n";
            try
            {
                auto invalid = missingImo.build();
                std::cout << "  ERROR: Invalid build succeeded (unexpected)\n";
            }
            catch (const std::invalid_argument& e)
            {
                std::cout << "  Correctly rejected: " << e.what() << "\n";
            }
        }

        std::cout << "\n";
    }

    {
        std::cout << "9. UniversalId: Real-world vessel examples\n";
        std::cout << "---------------------------------------------\n";

        const auto& vis = VIS::instance();
        const auto& gmod = vis.gmod(vis.latest());
        const auto& locations = vis.locations(vis.latest());
        const auto& codebooks = vis.codebooks(vis.latest());

        struct VesselSensor
        {
            std::string vesselName;
            int imoNumber;
            std::string sensorPath;
            std::string quantity;
        };

        std::vector<VesselSensor> sensors = { { "MV Queen Mary 2", 9241061, "411.1/C101.31-2", "temperature" },
                                              { "MS Oasis of the Seas", 9383936, "621.21/S90", "mass" },
                                              { "MS Symphony of the Seas", 9744001, "612.21/C701.23", "pressure" } };

        for (const auto& sensor : sensors)
        {
            auto imo = ImoNumber{ sensor.imoNumber };
            auto path = GmodPath::fromShortPath(sensor.sensorPath, gmod, locations);
            auto tag = codebooks[CodebookName::Quantity].createTag(sensor.quantity);

            if (path.has_value() && tag.has_value())
            {
                auto localIdBuilder = LocalIdBuilder::create(vis.latest()).withPrimaryItem(*path).withMetadataTag(*tag);

                auto universalId =
                    UniversalIdBuilder::create(vis.latest()).withImoNumber(imo).withLocalId(localIdBuilder).build();

                std::cout << sensor.vesselName << ":\n";
                std::cout << "  " << universalId.toString() << "\n\n";
            }
        }
    }

    {
        std::cout << "10. UniversalIdBuilder: Remove and add operations\n";
        std::cout << "---------------------------------------------------\n";

        auto original =
            UniversalId::fromString("data.dnv.com/IMO1234567/dnv-v2/vis-3-4a/411.1/C101.31-2/meta/qty-temperature");

        if (original.has_value())
        {
            std::cout << "Original: " << original->toString() << "\n";

            // Remove IMO number
            auto withoutImo = original->builder().withoutImoNumber();
            std::cout << "\nAfter withoutImoNumber():\n";
            std::cout << "  Is valid: " << std::boolalpha << withoutImo.isValid() << "\n";

            // Remove LocalId
            auto withoutLocalId = original->builder().withoutLocalId();
            std::cout << "\nAfter withoutLocalId():\n";
            std::cout << "  Is valid: " << std::boolalpha << withoutLocalId.isValid() << "\n";

            // Add them back
            const auto& vis = VIS::instance();
            const auto& gmod = vis.gmod(vis.latest());
            const auto& locations = vis.locations(vis.latest());
            const auto& codebooks = vis.codebooks(vis.latest());

            auto newImo = ImoNumber{ 9074729 };
            auto path = GmodPath::fromShortPath("621.21/S90", gmod, locations);
            auto tag = codebooks[CodebookName::Quantity].createTag("mass");

            if (path.has_value() && tag.has_value())
            {
                auto newLocalId = LocalIdBuilder::create(vis.latest()).withPrimaryItem(*path).withMetadataTag(*tag);
                auto rebuilt = withoutImo.withImoNumber(newImo).withLocalId(newLocalId).build();
                std::cout << "\nRebuilt with new components: " << rebuilt.toString() << "\n";
            }
        }

        std::cout << "\n";
    }

    return 0;
}

File: bindings/rust/vista-sdk/examples/core/universal_id.rs
use vista_sdk::core::codebook_name::CodebookName;
use vista_sdk::core::gmod_path::OwnedGmodPath;
use vista_sdk::core::imo_number::ImoNumber;
use vista_sdk::core::local_id_builder::OwnedLocalIdBuilder;
use vista_sdk::core::universal_id::UniversalId;
use vista_sdk::core::universal_id_builder::OwnedUniversalIdBuilder;
use vista_sdk::core::vis::Vis;

fn main() {
    println!("=== vista-sdk UniversalId Sample ===\n");

    {
        println!("1. UniversalIdBuilder: Building a simple UniversalId");
        println!("------------------------------------------------------");

        let vis = Vis::instance();
        let gmod = vis
            .gmod(vis.latest())
            .expect("latest() should always be a valid VIS version");
        let locations = vis
            .locations(vis.latest())
            .expect("latest() should always be a valid VIS version");
        let codebooks = vis
            .codebooks(vis.latest())
            .expect("latest() should always be a valid VIS version");

        let imo_number = ImoNumber::create(1234567).expect("1234567 should be a valid IMO number");

        let primary_item = OwnedGmodPath::from_short_path("411.1/C101.31-2", gmod, locations);
        let qty_tag = codebooks[CodebookName::Quantity].create_tag("temperature");

        if let (Some(primary_item), Some(qty_tag)) = (primary_item, qty_tag) {
            let local_id_builder = OwnedLocalIdBuilder::create(vis.latest())
                .with_primary_item(&primary_item)
                .with_metadata_tag(&qty_tag);

            let universal_id = OwnedUniversalIdBuilder::create(vis.latest())
                .with_imo_number(&imo_number)
                .with_local_id(&local_id_builder)
                .build()
                .expect("builder with IMO number and LocalIdBuilder should be valid");

            println!("Built UniversalId: {}", universal_id);
            println!("  IMO Number: {}", universal_id.imo_number());
            println!("  LocalId   : {}", universal_id.local_id());
        }

        println!();
    }

    {
        println!("2. UniversalIdBuilder: Building with full LocalId");
        println!("---------------------------------------------------");

        let vis = Vis::instance();
        let gmod = vis
            .gmod(vis.latest())
            .expect("latest() should always be a valid VIS version");
        let locations = vis
            .locations(vis.latest())
            .expect("latest() should always be a valid VIS version");
        let codebooks = vis
            .codebooks(vis.latest())
            .expect("latest() should always be a valid VIS version");

        let primary_item = OwnedGmodPath::from_short_path("621.21/S90", gmod, locations);
        let secondary_item = OwnedGmodPath::from_short_path("411.1/C101", gmod, locations);
        let qty_tag = codebooks[CodebookName::Quantity].create_tag("mass");
        let cnt_tag = codebooks[CodebookName::Content].create_tag("fuel.oil");
        let pos_tag = codebooks[CodebookName::Position].create_tag("inlet");

        if let (
            Some(primary_item),
            Some(secondary_item),
            Some(qty_tag),
            Some(cnt_tag),
            Some(pos_tag),
        ) = (primary_item, secondary_item, qty_tag, cnt_tag, pos_tag)
        {
            let local_id_builder = OwnedLocalIdBuilder::create(vis.latest())
                .with_primary_item(&primary_item)
                .with_secondary_item(&secondary_item)
                .with_metadata_tag(&qty_tag)
                .with_metadata_tag(&cnt_tag)
                .with_metadata_tag(&pos_tag);

            let imo_number = ImoNumber::create_from_str("IMO9074729")
                .expect("IMO9074729 should be a valid IMO number");
            let universal_id = OwnedUniversalIdBuilder::create(vis.latest())
                .with_imo_number(&imo_number)
                .with_local_id(&local_id_builder)
                .build()
                .expect("builder with IMO number and LocalIdBuilder should be valid");

            println!("Built UniversalId: {}", universal_id);
            println!("  Format       : {{naming-entity}}/{{imo-number}}{{local-id}}");
            println!("  Naming entity: {}", UniversalId::naming_entity());
            println!("  IMO Number   : {}", universal_id.imo_number());
        }

        println!();
    }

    {
        println!("3. UniversalId: Parsing from string");
        println!("-------------------------------------");

        let universal_id_str =
            "data.dnv.com/IMO1234567/dnv-v2/vis-3-4a/411.1/C101.31-2/meta/qty-temperature";

        let universal_id = OwnedUniversalIdBuilder::from_string(universal_id_str);

        if let Some(universal_id) = universal_id {
            println!("Parsed successfully: {}", universal_id);
            println!("  IMO Number : {}", universal_id.imo_number());
            println!("  LocalId    : {}", universal_id.local_id());
            println!("  VIS Version: {}", universal_id.local_id().version());
        } else {
            println!("Parse failed");
        }

        println!();
    }

    {
        println!("4. UniversalId: Parsing with error handling");
        println!("----------------------------------------------");

        let test_strings = [
            "data.dnv.com/IMO1234567/dnv-v2/vis-3-4a/411.1/C101.31-2/meta/qty-temperature",
            "data.dnv.com/IMO9074729/dnv-v2/vis-3-4a/621.21/S90/sec/411.1/C101/meta/qty-mass/cnt-fuel.oil/pos-inlet",
            "data.dnv.com/IMO9785811/dnv-v2/vis-3-7a/612.21/C701.23/C633/meta/calc~accumulate",
            "",
            "data.dnv.com/INVALID/dnv-v2/vis-3-4a/411.1/meta/qty-temperature",
            "wrong.entity/IMO1234567/dnv-v2/vis-3-4a/411.1/meta/qty-temperature",
            "data.dnv.com/IMO1234567/dnv-v2/vis-3-4a/INVALID/meta/qty-temperature",
        ];

        for str in test_strings {
            let (universal_id, errors) = OwnedUniversalIdBuilder::from_string_with_errors(str);

            let display = if str.is_empty() {
                "(empty)".to_string()
            } else if str.len() > 60 {
                format!("{}...", &str[..60])
            } else {
                str.to_string()
            };
            println!("Parsing: \"{}\"", display);

            if let Some(universal_id) = universal_id {
                println!("  Success");
                println!("    IMO: {}", universal_id.imo_number());
            } else {
                println!("  Failed:");
                if errors.has_errors() {
                    for error in errors.iter() {
                        println!("    - {}", error.message);
                    }
                }
            }
            println!();
        }
    }

    {
        println!("5. UniversalIdBuilder: Modifying existing UniversalIds");
        println!("--------------------------------------------------------");

        let original = OwnedUniversalIdBuilder::from_string(
            "data.dnv.com/IMO1234567/dnv-v2/vis-3-4a/411.1/C101.31-2/meta/qty-temperature",
        );

        if let Some(original) = original {
            println!("Original                    : {}", original);

            let vis = Vis::instance();
            let codebooks = vis
                .codebooks(vis.latest())
                .expect("latest() should always be a valid VIS version");

            let new_imo = ImoNumber::create(9074729).expect("9074729 should be a valid IMO number");
            let modified = original
                .builder()
                .with_imo_number(&new_imo)
                .build()
                .expect("builder with new IMO number should be valid");

            println!("Modified (new IMO)          : {}", modified);

            let cnt_tag = codebooks[CodebookName::Content].create_tag("water");
            if let Some(cnt_tag) = cnt_tag {
                let new_local_id_builder =
                    original.local_id().builder().with_metadata_tag(&cnt_tag);
                let modified2 = original
                    .builder()
                    .with_local_id(&new_local_id_builder)
                    .build()
                    .expect("builder with modified LocalIdBuilder should be valid");

                println!("Modified (added content tag): {}", modified2);
            }
        }

        println!();
    }

    {
        println!("6. UniversalId: Accessing components");
        println!("--------------------------------------");

        let universal_id = OwnedUniversalIdBuilder::from_string(
            "data.dnv.com/IMO9074729/dnv-v2/vis-3-4a/621.21/S90/sec/411.1/C101/meta/qty-mass/cnt-fuel.oil/pos-inlet",
        );

        if let Some(universal_id) = universal_id {
            println!("UniversalId: {}\n", universal_id);

            println!("IMO Number:");
            println!("  Value: {}", universal_id.imo_number());

            println!("\nLocalId components:");
            let local_id = universal_id.local_id();
            println!("  VIS Version   : {}", local_id.version());
            println!("  Primary item  : {}", local_id.primary_item());

            if let Some(secondary_item) = local_id.secondary_item() {
                println!("  Secondary item: {}", secondary_item);
            }

            println!("\nMetadata tags:");
            if let Some(quantity) = local_id.quantity() {
                println!("  Quantity: {}", quantity.value());
            }
            if let Some(content) = local_id.content() {
                println!("  Content : {}", content.value());
            }
            if let Some(position) = local_id.position() {
                println!("  Position: {}", position.value());
            }
        }

        println!();
    }

    {
        println!("7. UniversalId: Equality and comparison");
        println!("-----------------------------------------");

        let uid1 = OwnedUniversalIdBuilder::from_string(
            "data.dnv.com/IMO1234567/dnv-v2/vis-3-4a/411.1/C101.31-2/meta/qty-temperature",
        );
        let uid2 = OwnedUniversalIdBuilder::from_string(
            "data.dnv.com/IMO1234567/dnv-v2/vis-3-4a/411.1/C101.31-2/meta/qty-temperature",
        );
        let uid3 = OwnedUniversalIdBuilder::from_string(
            "data.dnv.com/IMO9074729/dnv-v2/vis-3-4a/411.1/C101.31-2/meta/qty-temperature",
        );
        let uid4 = OwnedUniversalIdBuilder::from_string(
            "data.dnv.com/IMO1234567/dnv-v2/vis-3-4a/411.1/C101.31-2/meta/qty-pressure",
        );

        if let (Some(uid1), Some(uid2), Some(uid3), Some(uid4)) = (uid1, uid2, uid3, uid4) {
            println!("UniversalId1: Same IMO, same LocalId");
            println!("UniversalId2: Same IMO, same LocalId");
            println!("UniversalId3: Different IMO, same LocalId");
            println!("UniversalId4: Same IMO, different LocalId\n");

            println!("uid1 == uid2? {} (same IMO & LocalId)", *uid1 == *uid2);
            println!("uid1 == uid3? {} (different IMO)", *uid1 == *uid3);
            println!("uid1 == uid4? {} (different LocalId)", *uid1 == *uid4);
            println!("uid1 != uid3? {}", *uid1 != *uid3);
        }

        println!();
    }

    {
        println!("8. UniversalIdBuilder: Validation");
        println!("-----------------------------------");

        let vis = Vis::instance();
        let gmod = vis
            .gmod(vis.latest())
            .expect("latest() should always be a valid VIS version");
        let locations = vis
            .locations(vis.latest())
            .expect("latest() should always be a valid VIS version");
        let codebooks = vis
            .codebooks(vis.latest())
            .expect("latest() should always be a valid VIS version");

        let imo_number = ImoNumber::create(1234567).expect("1234567 should be a valid IMO number");
        let primary_item = OwnedGmodPath::from_short_path("411.1/C101.31-2", gmod, locations);
        let qty_tag = codebooks[CodebookName::Quantity].create_tag("temperature");

        if let (Some(primary_item), Some(qty_tag)) = (primary_item, qty_tag) {
            let local_id_builder = OwnedLocalIdBuilder::create(vis.latest())
                .with_primary_item(&primary_item)
                .with_metadata_tag(&qty_tag);

            let valid_builder = OwnedUniversalIdBuilder::create(vis.latest())
                .with_imo_number(&imo_number)
                .with_local_id(&local_id_builder);

            println!("Valid builder:");
            println!("  Is valid: {}", valid_builder.is_valid());

            let missing_imo =
                OwnedUniversalIdBuilder::create(vis.latest()).with_local_id(&local_id_builder);
            println!("\nBuilder missing IMO number:");
            println!("  Is valid: {}", missing_imo.is_valid());

            let missing_local_id =
                OwnedUniversalIdBuilder::create(vis.latest()).with_imo_number(&imo_number);
            println!("\nBuilder missing LocalId:");
            println!("  Is valid: {}", missing_local_id.is_valid());

            println!("\nAttempting to build without IMO number:");
            match missing_imo.build() {
                Ok(_) => println!("  ERROR: Invalid build succeeded (unexpected)"),
                Err(error) => println!("  Correctly rejected: {}", error),
            }
        }

        println!();
    }

    {
        println!("9. UniversalId: Real-world vessel examples");
        println!("---------------------------------------------");

        let vis = Vis::instance();
        let gmod = vis
            .gmod(vis.latest())
            .expect("latest() should always be a valid VIS version");
        let locations = vis
            .locations(vis.latest())
            .expect("latest() should always be a valid VIS version");
        let codebooks = vis
            .codebooks(vis.latest())
            .expect("latest() should always be a valid VIS version");

        let sensors = [
            ("MV Queen Mary 2", 9241061, "411.1/C101.31-2", "temperature"),
            ("MS Oasis of the Seas", 9383936, "621.21/S90", "mass"),
            (
                "MS Symphony of the Seas",
                9744001,
                "612.21/C701.23",
                "pressure",
            ),
        ];

        for (vessel_name, imo_number, sensor_path, quantity) in sensors {
            let imo = ImoNumber::create(imo_number).expect("valid IMO number");
            let path = OwnedGmodPath::from_short_path(sensor_path, gmod, locations);
            let tag = codebooks[CodebookName::Quantity].create_tag(quantity);

            if let (Some(path), Some(tag)) = (path, tag) {
                let local_id_builder = OwnedLocalIdBuilder::create(vis.latest())
                    .with_primary_item(&path)
                    .with_metadata_tag(&tag);

                let universal_id = OwnedUniversalIdBuilder::create(vis.latest())
                    .with_imo_number(&imo)
                    .with_local_id(&local_id_builder)
                    .build()
                    .expect("builder with IMO number and LocalIdBuilder should be valid");

                println!("{}:", vessel_name);
                println!("  {}\n", universal_id);
            }
        }
    }

    {
        println!("10. UniversalIdBuilder: Remove and add operations");
        println!("---------------------------------------------------");

        let original = OwnedUniversalIdBuilder::from_string(
            "data.dnv.com/IMO1234567/dnv-v2/vis-3-4a/411.1/C101.31-2/meta/qty-temperature",
        );

        if let Some(original) = original {
            println!("Original: {}", original);

            let without_imo = original.builder().without_imo_number();
            println!("\nAfter without_imo_number():");
            println!("  Is valid: {}", without_imo.is_valid());

            let without_local_id = original.builder().without_local_id();
            println!("\nAfter without_local_id():");
            println!("  Is valid: {}", without_local_id.is_valid());

            let vis = Vis::instance();
            let gmod = vis
                .gmod(vis.latest())
                .expect("latest() should always be a valid VIS version");
            let locations = vis
                .locations(vis.latest())
                .expect("latest() should always be a valid VIS version");
            let codebooks = vis
                .codebooks(vis.latest())
                .expect("latest() should always be a valid VIS version");

            let new_imo = ImoNumber::create(9074729).expect("9074729 should be a valid IMO number");
            let path = OwnedGmodPath::from_short_path("621.21/S90", gmod, locations);
            let tag = codebooks[CodebookName::Quantity].create_tag("mass");

            if let (Some(path), Some(tag)) = (path, tag) {
                let new_local_id_builder = OwnedLocalIdBuilder::create(vis.latest())
                    .with_primary_item(&path)
                    .with_metadata_tag(&tag);
                let rebuilt = without_imo
                    .with_imo_number(&new_imo)
                    .with_local_id(&new_local_id_builder)
                    .build()
                    .expect("builder with new IMO number and LocalIdBuilder should be valid");

                println!("\nRebuilt with new components: {}", rebuilt);
            }
        }

        println!();
    }
}
